### ISSUE TITLE
BE/supportability: Reliable Orc8r upgrade path

### ISSUE THREAD

**Issue by talkhasib** [magma/magma#7861]
When partner has 1000 AGWs in the field, distributed across 100 customers. There is a new feature that we really want on the ORC8R. You verify that the upgrade works. We also verify it using our dev ORC8R. But then, while upgrading production ORC8R, something goes wrong and ORC8R crashes so badly that we have to build a new one. We can build a new one quickly, that’s not an issue, but all 1000 AGWs in the field have certificates based on the now defunct ORC8R. How do we ensure that 1000 AGWs in the field do not have to be reinstalled with the new ORC8R certificates? Having common root CA is one possible approach. Backing up certificate and the private key from the ORC8R so they can be copied to a new ORC8R if needed is another. There may be other options – just tell us what works.
There is an even bigger problem: what happens with all the customer data provisioned on the old ORC8R? Can the new ORC8R access the same database as the old one (my assumption is that the databases are not corrupted)?

To recap: with the current architecture, ORC8R is a single point of failure. While we can live with this at the moment, as the number of AGWs in the field increases, we have to understand what to do if ORC8R fails.

**Comment by amarpad** [comment]
I think it is beyond just the cert. For the FreedomFi upgrade to 1.5 we lost 100+ AGWs when it went headless due to the tenant name issue. So I will our current approach of teraforming upgrades is not really k8s friendly. Also note for the cases where we are using FeG like inbound roaming, orc8r going out to lunch affects signaling.

**Comment by arunuke** [comment]
The current terraform upgrade process replaces existing pods with new pods and removes the old pods so that the new "state" is as expected. What is needed (and has been talked about a number of times before and is a proposal currently) is the blue/green model where both clusters exists within the same footprint and the green (new) cluster is tested and validated and is used for all new connections before the blue (old) cluster is shut down after the last of old connections is closed (or forcefully migrated after a certain Continuous Hours of Operation).

The ongoing discussions are around what tools to use and the following key questions.

(1) using k8s natively is an option, but orc8r is complex enough that having helm to package different "applications" has been used for the deploy. Should we standardize on k8s since it is the lowest common denominator or instead use helm.

(2) aws/cloud resources are deployed using terraform and hence are also upgraded via terraform (ex: a db upgrade for rds). Considering this, there is an effort ongoing to achieve blue/green via terraform upgrades.

**Comment by ekowtaylor** [comment]
@arunuke please add story points to this issue
